### PR TITLE
bugfix for #277 : load_results properly handles experiments dtypes

### LINKED ISSUE #277: Visual Analysis Example throws an error
https://github.com/quaquel/EMAworkbench/blob/b8748d9db912b80a25cd3f25b0f309187fb13dc5/docs/source/indepth_tutorial/open-exploration.ipynb

Running the example code for "Visual Analysis" provided in the extended tutorial for "Open Exploration" produces the following error:

```
  File ".../python3.11/site-packages/ema_workbench/analysis/plotting_util.py", line 646, in make_continuous_grouping_specifiers
    step = (maximum - minimum) / nr_of_groups
           ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
TypeError: unsupported operand type(s) for /: 'int' and 'NoneType'
```

Following the call-stack, it seems like if ever  `pairs_plotting.pairs_scatter` is called with a `group_by` but not a `grouping_specifiers`, then `grouping_specifiers` defaults to None, and you will get the above TypeError. The error won't occur when the `policy` (or whatever column is provided) has `dtype=category`, but it seems that is not the dtype of `policy`. I might be wrong on this and there might be some cases where it works, but it doesn't seem like it to me. 

**INFO:** I am running this code after loading saved results using `load_results()`. It could be that the real issue is that the dtype is lost in this process.

~~**Suggestion**: Add `grouping_specifiers = range(policies)` to the `pairs_scatter` call in the example for robustness.~~

EDIT: removed suggestion because I think that would cause errors when the dtype is correct. See discussion below; issue is in the dtype changing when `save/load_results` are used.

### PR DESCRIPTION
**Pull request by quaquel**
Bugfix for load_results which fixes #277. In parsing experiments, the dtype metadata is now correctly applied.  It also expands save_results unit test to cover all dtypes used in storing experiments.

### CODE REVIEW

File: ema_workbench/util/utilities.py [hunk 1]
```diff
@@ -64,6 +64,9 @@ def load_results(file_name):
             except TypeError:
                 dtype = pd.api.types.pandas_dtype(dtype)
 
+            if experiments[name].dtype is not dtype:
+                experiments[name] = experiments[name].astype(dtype)
```
Comment (EwoutH) [inline ema_workbench/util/utilities.py:68]: Can we add a log/print/warning statement here that notes that there was a data type conversion? That might make it possible to identify future bugs or unintended behaviour.

Something like:
```Python
f"Data type of {name} was incorrectly loaded as {experiments[name].dtype}, converting back to {dtype}."
```

Comment (quaquel) [reply ema_workbench/util/utilities.py:68]: That won't really work. We rely on pandas.read_csv to load the csv. This does a best guess dtype. But if you have e.g., integers that refer to a category in a categorical uncertainty (as happens with e.g., vensim models quite often), read_csv will load this as an integer dtype which then is converted into a categorical dtype (for correct handling in e.g. PRIM this is quite relevant).

File: test/test_util/test_utilities.py [hunk 2]
```diff
@@ -85,19 +87,19 @@ def test_save_results(self):
...
-            index=np.arange(nr_experiments), columns={"x": float, "y": float}
+        cases = np.empty(
+            nr_experiments, dtype=[("x", float), ("y", int), ("z", bool), ("q", object)]
         )
+        experiments = pd.DataFrame.from_records(cases)
 
         experiments["x"] = np.random.rand(nr_experiments)
-        experiments["y"] = np.random.rand(nr_experiments)
+        experiments["y"] = np.random.randint(0, 10, size=nr_experiments)
+        experiments["y"] = np.random.randint(0, 1, size=nr_experiments, dtype=bool)
```
Comment (EwoutH) [inline test/test_util/test_utilities.py:100]: This line seems to overwrite the previous value of `experiments["y"]` in the line before (it's assigned twice, so the first is overwritten). Is that intended?

File: ema_workbench/util/utilities.py (hunk 1 above)
Comment (EwoutH) [reply ema_workbench/util/utilities.py:68]: Right, thanks for the explanation.

**Review by EwoutH — APPROVED** [review]
One last comment on the test, the rest looks good, thanks!

Do you want this into 2.4.1?

File: test/test_util/test_utilities.py (hunk 2 above)
Comment (quaquel) [reply test/test_util/test_utilities.py]: should be "z". See extra commit

**Pull request merged**